cli/compose/loader/loader.go

```diff
@@ -352,20 +357,34 @@ func resolveEnvironment(serviceConfig *types.ServiceConfig, workingDir string) e
 		}
 
 		for k, v := range runconfigopts.ConvertKVStringsToMap(envVars) {
-			environment[k] = v
+			interpolatedV, ok := lookupEnv(k)
+			if ok {
+				// lookupEnv is prioritized over the file
+				environment[k] = interpolatedV
+			} else {
+				environment[k] = v
+			}
 		}
 	}
 
 	for k, v := range serviceConfig.Environment {
-		environment[k] = v
+		if v == "" {
+			interpolatedV, ok := lookupEnv(k)
+			if ok {
+				environment[k] = interpolatedV
+			} else {
+				delete(environment, k)
```

Comment:
Yes, you are right, I didn't know that 😅 

Opened https://github.com/docker/docker/pull/31634 for clarifying this behavior in the doc.

I'll update this PR as well in short.